v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,497 @@
+    {
+        if (state == PROVIDER_STATE_NOT_REGISTERED)
+        {
+            /* Codes_SRS_LOG_SINK_ETW_01_006: [ log_sink_etw_log shall register the ETW TraceLogging provider by calling TraceLoggingRegister (TraceLoggingRegister_EventRegister_EventSetInformation). ]*/
+            TLG_STATUS register_result = TraceLoggingRegister(g_my_component_provider);
+            if (SUCCEEDED(register_result))
+            {
+                /* Codes_SRS_LOG_SINK_ETW_01_007: [ log_sink_etw_log shall switch the state to REGISTERED. ]*/
+                (void)InterlockedExchange(&etw_provider_state, PROVIDER_STATE_REGISTERED);
+                /* Codes_SRS_LOG_SINK_ETW_01_008: [ log_sink_etw_log shall printf the fact that the provider was registered and from which executable (as obtained by calling _pgmptr). ]*/
+
+                char* executable_full_path_name;
+                if (_get_pgmptr(&executable_full_path_name) != 0)
+                {
+                    /* Codes_SRS_LOG_SINK_ETW_01_083: [ If _get_pgmptr fails, the executable shall be printed as UNKNOWN. ]*/
+                    executable_full_path_name = "UNKNOWN";
+                }
+
+                /* Codes_SRS_LOG_SINK_ETW_01_008: [ log_sink_etw_log shall emit a LOG_LEVEL_INFO event as a self test , printing the fact that the provider was registered and from which executable (as obtained by calling _get_pgmptr). ]*/
+                log_sink_etw_log(LOG_LEVEL_INFO, NULL, __FILE__, __FUNCTION__, __LINE__, "ETW provider was registered succesfully (self test). Executable file full path name = %s", MU_P_OR_NULL(executable_full_path_name));
+
+                break;
+            }
+            else
+            {
+                (void)printf("ETW provider was NOT registered.\r\n");
+                (void)InterlockedExchange(&etw_provider_state, PROVIDER_STATE_NOT_REGISTERED);
+            }
+        }
+    }
+}
+
+// want these aligned to 1 byte (copy/pasted from Tracelogging code)
+__pragma(pack(push, 1))
+typedef struct SELF_DESCRIBED_EVENT_TAG
+{
+    uint8_t _tlgBlobTyp;
+    uint8_t _tlgChannel;
+    uint8_t _tlgLevel;
+    uint8_t _tlgOpcode;
+    uint64_t _tlgKeyword;
+    uint16_t _tlgEvtMetaSize;
+    uint16_t _tlgEvtTag;
+    uint8_t metadata[];
+} SELF_DESCRIBED_EVENT;
+__pragma(pack(pop))
+
+// This function was written with a little bit of reverse engineering of TraceLogging and guidance from 
+// the TraceLogging.h header about the format of the self described events
+static void internal_emit_self_described_event(const char* event_name, uint16_t event_name_length, uint8_t trace_level, const LOG_CONTEXT_PROPERTY_VALUE_PAIR* context_property_value_pairs, uint32_t property_value_count, const char* message, const char* file, const char* func, int32_t line, va_list args)
+{
+    TraceLoggingHProvider const _tlgProv = g_my_component_provider;
+    if (trace_level < _tlgProv->LevelPlus1)
```

Comment:
>LevelPlus1 [](http://example.com/codeflow?start=32&length=10)

What makes this value change? :)

Reply:
The guts of Tracelogging. I did check it takes effect by running the perf tests. And I plan some INT tests for it in the next PRs.